mono/mini/mini-posix.c

```diff
@@ -1117,6 +1117,11 @@ mono_post_native_crash_handler (const char *signal, MonoContext *mctx, MONO_SIG_
 		abort ();
 #endif
 	}
+	mono_chain_signal (info->si_signo, info, context);
+
+	// we remove Mono's signal handlers from crashing signals in mono_handle_native_crash(), so re-raising will now allow the OS to handle the crash
+	// TODO: perhaps we can always use this to abort, instead of explicit exit()/abort() as we do above
+	raise (info->si_signo);
```

Comment:
Is it possible to:

1. Restore the default signal action
2. Restore the original context (i.e. restore the registers, etc. to the values when the signal was raised), so that the process resumes execution with the default signal action (which will presumably terminate the process and cause the OS to produce a crash report).

The point is to leave the process in the same state it was when the signal was raised, except without Mono's signal handler. The reason for this is that when there's a crash in native code, it's quite often very difficult to figure out what happened, and having the original state can be helpful. In particular with crashes due to bugs in Apple's software, it would be much better to be able to provide Apple engineers with a crash report that accurately reflects the process state at the time of the crash instead of a process where Mono stomped all over the place.

Reply:
@rolf (1) is done (now more completely) in `mono_handle_native_crash` 

(2) I will have to think about how that would be possible - restoring the context would naturally divert execution to the new (old) IP.

Reply:
@Rolf On further thought, just restoring the context might work - without advancing the IP, the same instruction will execute as before, presumably inducing the same signal which then would be handled by the OS instead of Mono.